components/places/src/storage.rs

```diff
@@ -263,21 +270,311 @@ fn add_visit(db: &impl ConnExt,
+                                                    |row| HistoryRecordVisit {
+                date: row.get::<_, Timestamp>("date").into(),
+                transition: row.get::<_, u8>("transition"),
+            })?;
+            let visits = visit_rows.collect::<RusqliteResult<Vec<_>>>()?;
+            if result.contains_key(&page.guid) {
+                // should be impossible!
+                warn!("Found {:?} in both tombstones and live records", &page.guid);
+                continue;
+            }
+            if visits.len() == 0 {
+                info!("Page {:?} is flagged to be uploaded, but has no visits - skipping", &page.guid);
+                continue;
+            }
+            trace!("outgoing record {:?}", &page.guid);
+            ids_to_update.push(page.row_id);
+            db.execute_named_cached(insert_meta_sql,
+                                    &[(":row_id", &page.row_id),
+                                      (":change_delta", &page.sync_change_counter)])?;
+
+            result.insert(page.guid.clone(), OutgoingInfo::Record(HistoryRecord {
+                id: page.guid,
+                title: page.title,
+                hist_uri: page.url.to_string(),
+                sortindex: page.frecency,
+                ttl: HISTORY_TTL,
+                visits,
+            }));
+        }
+
+        // We need to update the sync status of these items now rather than after
+        // the upload, because if we are interrupted between upload and writing
+        // we could end up with local items with state New even though we
+        // uploaded them.
+        sql_support::each_chunk(&ids_to_update, |chunk, _| -> Result<()> {
+            db.conn().execute(&format!("UPDATE moz_places SET sync_status={status}
+                                 WHERE id IN ({vars})",
+                                vars = sql_support::repeat_sql_vars(chunk.len()),
+                                status = SyncStatus::Normal as u8),
+                              chunk)?;
+            Ok(())
+        })?;
+
+        Ok(result)
+    }
+
+    pub fn finish_outgoing(db: &Connection) -> Result<()> {
+        // So all items *other* than those above must be set to "not dirty"
+        // (ie, status=SyncStatus::Normal, change_counter=0). Otherwise every
+        // subsequent sync will continue to add more and more local visits
+        // until every visit we have is uploaded. And we only want to do it
+        // at the end of the sync because if we are interrupted, we'll end up
+        // thinking we have nothing to upload.
+        // BUT - this is potentially alot of rows! Because we want "NOT IN (...)"
+        // we can't do chunking and building a literal string with the ids seems
+        // wrong and likely to hit max sql length limits.
+        // So we use a temp table.
+        trace!("Updating all synced rows");
+        // XXX - is there a better way to express this SQL? Multi-selects
+        // doesn't seem ideal...
```

Comment:
We could abuse `INSERT...ON CONFLICT UPDATE`...try to insert the ID (and placeholder/default values for `NOT NULL` columns without a `DEFAULT`), let that fail, then update the change counter...but I think this is better.